src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java

```diff
@@ -481,19 +482,18 @@ public int hashCode() {
 
   /** Lookup the remote cache for the given {@link RemoteAction}. {@code null} if not found. */
   @Nullable
-  public RemoteActionResult lookupCache(RemoteAction action)
+  public CachedActionResult lookupCache(RemoteAction action)
       throws IOException, InterruptedException {
     checkState(shouldAcceptCachedResult(action.spawn), "spawn doesn't accept cached result");
 
-    ActionResult actionResult =
-        remoteCache.downloadActionResult(
-            action.remoteActionExecutionContext, action.actionKey, /* inlineOutErr= */ false);
+    CachedActionResult cachedActionResult = remoteCache.downloadActionResult(
+        action.remoteActionExecutionContext, action.actionKey, /* inlineOutErr= */ false);
 
-    if (actionResult == null) {
+    if (cachedActionResult == null) {
```

Comment:
We have lots of places checking for `cachedActionResult != null` and `cachedActionResult.actionResult() != null`. Please consider updating `downloadActionResult` to always return a future that will emit non-null `CachedActionResult` on success. We can then differentiate cache-hit and cache-miss by only checking `cachedActionResult.cacheResult() != null`.

Reply:
done